src/time_tracker/invoices/core.clj

```diff
@@ -1,62 +1,53 @@
 (ns time-tracker.invoices.core
   (:require [clojure.data.csv :as csv]
             [time-tracker.timers.core :as timers-core]
-            [clojure.algo.generic.functor :refer [fmap]]))
-
-(defn empty-time-map
-  "Returns a map of {user-id {project-id 0}}
-  for all users and projects. The last entry is the number of hours.
-  There must be at least one user and one project."
-  [users projects]
-  (->> (for [user-id (keys users)
-             project-id (keys projects)]
-         {user-id {project-id 0}})
-       (apply merge-with merge)))
+            [clojure.algo.generic.functor :refer [fmap]]
+            [time-tracker.util :as util]))
 
 (defn- seconds->hours
   [seconds]
   (double (/ seconds 3600)))
 
-(defn add-hours
-  "Adds the hours logged against the timer entry
-  to the time map."
-  [time-map {:keys [project-id app-user-id] :as timer}]
-  (update-in time-map
-             [app-user-id project-id]
-             #(+ % (seconds->hours (timers-core/elapsed-time timer)))))
-
-(defn build-time-map
-  "Returns a map of {user-id {project-id hours-logged}}.
-  There must be at least one user and one project."
-  [users projects timers]
-  (reduce #(add-hours %1 %2)
-          (empty-time-map users projects)
-          (vals timers)))
-
 (defn- round-to-places
   "Rounds a floating point number to `places` decimal places."
   [number places]
   (.setScale (bigdec number) places java.math.BigDecimal/ROUND_HALF_UP))
 
+(defn add-hours
+  "Adds the hours logged against the timer entry
+  to the `user-id->hours` map."
+  [user-id->hours {:keys [app-user-id] :as timer}]
+  (update user-id->hours app-user-id
+          (fn [current-hours]
+            (let [hours-to-add (-> (timers-core/elapsed-time timer)
+                                   (seconds->hours)
+                                   (round-to-places 4))]
+              (if (some? current-hours)
+                (+ current-hours hours-to-add)
+                hours-to-add)))))
+
+(defn build-user-id->hours
+  "Returns a map of {user-id hours-logged}"
+  [required-user-ids timers]
+  (let [user-id->zero (zipmap required-user-ids (repeat 0))]
+    (reduce add-hours user-id->zero (vals timers))))
```

Comment:
It's possible (but unlikely) that a user may not have logged any time, but their name still needs to appear in this invoice (maybe because they were staffed on the client). In this case there might not be any timers in their name at all. This is why the `required-user-ids` field is necessary.